modules/distributor/queue/queue.go

```diff
@@ -0,0 +1,224 @@
+		pushesTotalMetrics:         pushesTotalMetrics,
+		pushesFailuresTotalMetrics: pushesFailuresTotalMetric,
+		lengthMetric:               lengthMetric,
+		readOnly:                   atomic.NewBool(false),
+	}
+}
+
+// Push pushes data onto a queue.
+// If the queue is full, the data is dropped
+func (m *Queue[T]) Push(ctx context.Context, data T) error {
+	if m.readOnly.Load() {
+		return fmt.Errorf("queue is read-only")
+	}
+
+	m.pushesTotalMetrics.WithLabelValues(m.name, m.tenantID).Inc()
+
+	req := &request[T]{
+		data: data,
+	}
+
+	select {
+	case <-ctx.Done():
+		m.pushesFailuresTotalMetrics.WithLabelValues(m.name, m.tenantID).Inc()
+		return fmt.Errorf("failed to push data to queue for tenant=%s and queue_name=%s: %w", m.tenantID, m.name, ctx.Err())
+	default:
+	}
+
+	select {
+	case m.reqChan <- req:
+		m.lengthMetric.WithLabelValues(m.name, m.tenantID).Inc()
+		return nil
+	default:
+	}
+
+	// Fail fast if the queue is full
+	m.pushesFailuresTotalMetrics.WithLabelValues(m.name, m.tenantID).Inc()
+
+	return fmt.Errorf("failed to push data to queue for tenant=%s and queue_name=%s: queue is full", m.tenantID, m.name)
+}
+
+func (m *Queue[T]) StartWorkers() {
+	for i := 0; i < m.workerCount; i++ {
+		m.wg.Add(1)
+
+		go m.worker()
+	}
+}
+
+// Name returns queue name.
+func (m *Queue[T]) Name() string {
+	return m.name
+}
+
+// TenantID returns the tenant id.
+func (m *Queue[T]) TenantID() string {
+	return m.tenantID
+}
+
+// Size returns the size of the queue.
+func (m *Queue[T]) Size() int {
+	return m.size
+}
+
+// WorkerCount returns the number of expected workers.
+func (m *Queue[T]) WorkerCount() int {
+	return m.workerCount
+}
+
+// ShouldUpdate returns true if the queue size or worker count (alive or total) has changed
+func (m *Queue[T]) ShouldUpdate(size, workerCount int) bool {
+	return m.size != size || m.workerCount != workerCount
+}
+
+func (m *Queue[T]) Shutdown(ctx context.Context) error {
+	// Call to stopWorkers only once
+	if m.readOnly.CAS(false, true) {
+		return m.stopWorkers(ctx)
+	}
+
+	return nil
+}
+
+func (m *Queue[T]) worker() {
+	defer m.wg.Done()
+
+	for {
+		select {
+		case req := <-m.reqChan:
+			m.lengthMetric.WithLabelValues(m.name, m.tenantID).Dec()
+			m.forwardRequest(req)
+		case <-m.workersCloseCh:
+			// Forces to always trying to pull from the queue before exiting
+			// This is important during shutdown to ensure that the queue is drained
+			select {
+			case req, ok := <-m.reqChan:
+				if !ok { //closed
+					return
+				}
+
+				m.lengthMetric.WithLabelValues(m.name, m.tenantID).Dec()
+				m.forwardRequest(req)
+			default:
+				return
+			}
+		}
+	}
+}
+
+func (m *Queue[T]) forwardRequest(req *request[T]) {
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+
+	ctx = user.InjectOrgID(ctx, m.tenantID)
+
+	if err := m.fn(ctx, req.data); err != nil {
+		_ = level.Error(m.logger).Log("msg", "pushing with forwarder failed", "err", err)
```

Comment:
nit: i'd prefer that whatever passes in the forward function handles its own errors. i don't see a ton of value in returning an error and just logging it.